File: src/encoding.cpp
#include "encoding.hpp"

namespace causal_chess {

torch::Tensor board_to_tensor(const chess::Board& board) {
    // Create zero-initialized tensor
    auto options = torch::TensorOptions().dtype(torch::kFloat32);
    torch::Tensor tensor = torch::zeros({NUM_PLANES, 8, 8}, options);
    float* data = tensor.data_ptr<float>();

    // 1. Piece planes (0-11)
    for (int sq = 0; sq < 64; ++sq) {
        chess::Square square(sq);
        chess::Piece piece = board.at(square);
        if (piece != chess::Piece::NONE) {
            int type_idx = static_cast<int>(piece.type());
            int color_idx = static_cast<int>(piece.color());
            int plane_idx = color_idx * 6 + type_idx;

            int rank = square.rank(); // 0..7
            int file = square.file(); // 0..7

            // Index in flat 1D array for (plane, rank, file)
            // Shape is (15, 8, 8), so strides are (64, 8, 1)
            int index = plane_idx * 64 + rank * 8 + file;
            data[index] = 1.0f;
        }
    }

    // 2. Side to move plane (12)
    if (board.sideToMove() == chess::Color::WHITE) {
        for (int i = 0; i < 64; ++i) {
            data[12 * 64 + i] = 1.0f;
        }
    }

    // 3. Castling rights planes (13 & 14)
    auto cr = board.castlingRights();

    // White castling (plane 13)
    if (cr.has(chess::Color::WHITE, chess::Board::CastlingRights::Side::QUEEN_SIDE)) {
        // a1 (file 0, rank 0)
        data[13 * 64 + 0 * 8 + 0] = 1.0f;
    }
    if (cr.has(chess::Color::WHITE, chess::Board::CastlingRights::Side::KING_SIDE)) {
        // h1 (file 7, rank 0)
        data[13 * 64 + 0 * 8 + 7] = 1.0f;
    }

    // Black castling (plane 14)
    if (cr.has(chess::Color::BLACK, chess::Board::CastlingRights::Side::QUEEN_SIDE)) {
        // a8 (file 0, rank 7)
        data[14 * 64 + 7 * 8 + 0] = 1.0f;
    }
    if (cr.has(chess::Color::BLACK, chess::Board::CastlingRights::Side::KING_SIDE)) {
        // h8 (file 7, rank 7)
        data[14 * 64 + 7 * 8 + 7] = 1.0f;
    }

    return tensor;
}

torch::Tensor batch_boards_to_tensor(const std::vector<chess::Board>& boards) {
    std::vector<torch::Tensor> tensors;
    tensors.reserve(boards.size());
    for (const auto& board : boards) {
        tensors.push_back(board_to_tensor(board));
    }
    return torch::stack(tensors);
}

} // namespace causal_chess

File: src/causal_chess/encoding.py
"""Board encoding: converts chess.Board to a tensor for the value network.

Encoding uses 15 planes of 8x8, all values White-relative:
  Planes 0-5:   White pieces (P, N, B, R, Q, K) — binary
  Planes 6-11:  Black pieces (P, N, B, R, Q, K) — binary
  Plane 12:     Side to move (all 1s if White to move, all 0s if Black)
  Plane 13:     White castling rights (1s at a1 if queenside, h1 if kingside)
  Plane 14:     Black castling rights (1s at a8 if queenside, h8 if kingside)
"""

import chess
import numpy as np
import torch


# Mapping from (piece_type, color) to plane index
_PIECE_PLANE: dict[tuple[int, bool], int] = {
    (chess.PAWN, chess.WHITE): 0,
    (chess.KNIGHT, chess.WHITE): 1,
    (chess.BISHOP, chess.WHITE): 2,
    (chess.ROOK, chess.WHITE): 3,
    (chess.QUEEN, chess.WHITE): 4,
    (chess.KING, chess.WHITE): 5,
    (chess.PAWN, chess.BLACK): 6,
    (chess.KNIGHT, chess.BLACK): 7,
    (chess.BISHOP, chess.BLACK): 8,
    (chess.ROOK, chess.BLACK): 9,
    (chess.QUEEN, chess.BLACK): 10,
    (chess.KING, chess.BLACK): 11,
}

NUM_PLANES = 15


def board_to_tensor(board: chess.Board) -> torch.Tensor:
    """Encode a chess.Board as a (15, 8, 8) float32 tensor.

    The encoding is White-relative: plane layout is always from White's
    perspective regardless of whose turn it is.

    Args:
        board: The chess position to encode.

    Returns:
        A float32 tensor of shape (15, 8, 8).
    """
    planes = np.zeros((NUM_PLANES, 8, 8), dtype=np.float32)

    # Planes 0-11: piece positions
    for square, piece in board.piece_map().items():
        rank = square // 8  # 0 = rank 1, 7 = rank 8
        file = square % 8   # 0 = a-file, 7 = h-file
        plane_idx = _PIECE_PLANE[(piece.piece_type, piece.color)]
        planes[plane_idx, rank, file] = 1.0

    # Plane 12: side to move
    if board.turn == chess.WHITE:
        planes[12, :, :] = 1.0

    # Plane 13: White castling rights
    if board.has_queenside_castling_rights(chess.WHITE):
        planes[13, 0, 0] = 1.0  # a1
    if board.has_kingside_castling_rights(chess.WHITE):
        planes[13, 0, 7] = 1.0  # h1

    # Plane 14: Black castling rights
    if board.has_queenside_castling_rights(chess.BLACK):
        planes[14, 7, 0] = 1.0  # a8
    if board.has_kingside_castling_rights(chess.BLACK):
        planes[14, 7, 7] = 1.0  # h8

    return torch.from_numpy(planes)


def batch_boards_to_tensor(boards: list[chess.Board]) -> torch.Tensor:
    """Encode multiple boards into a batched tensor.

    Args:
        boards: List of chess positions to encode.

    Returns:
        A float32 tensor of shape (len(boards), 15, 8, 8).
    """
    return torch.stack([board_to_tensor(b) for b in boards])
